### ISSUE TITLE
ContentTypes...

### ISSUE THREAD

**Issue by olasitarska** [potatolondon/wagtail-gae#3]
Traceback at the end... but basically...

Wagtail uses ContentTypes a lot. For everything. ContentType is even a ForeignKey on `Page` model, which is what everything is based on. I think what happens is that during migrations, we create some ContentType objects for some Page objects, and this happens before our ContentType patch kicks in. They're created in Django's ContentType model (is that even possible?!) and they later don't exist in Djangae's ContentType model (is that a thing?) 

I need to dig in more into what Djangae even does with ContentTypes and why I get DoesNotExist everywhere.

```
ERROR    2016-08-05 17:13:19,033 base.py:284] Internal Server Error: /admin/pages/
Traceback (most recent call last):
  File "/Users/ositarska/playground/mysite/sitepackages/django/core/handlers/base.py", line 149, in get_response
    response = self.process_exception_by_middleware(e, request)
  File "/Users/ositarska/playground/mysite/sitepackages/django/core/handlers/base.py", line 147, in get_response
    response = wrapped_callback(request, *callback_args, **callback_kwargs)
  File "/Users/ositarska/playground/mysite/sitepackages/django/views/decorators/cache.py", line 43, in _cache_controlled
    response = viewfunc(request, *args, **kw)
  File "/Users/ositarska/playground/mysite/sitepackages/wagtail/wagtailadmin/decorators.py", line 23, in decorated_view
    return view_func(request, *args, **kwargs)
  File "/Users/ositarska/playground/mysite/sitepackages/wagtail/wagtailadmin/views/pages.py", line 103, in index
    'do_paginate': do_paginate,
  File "/Users/ositarska/playground/mysite/sitepackages/django/shortcuts.py", line 67, in render
    template_name, context, request=request, using=using)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/loader.py", line 97, in render_to_string
    return template.render(context, request)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/backends/django.py", line 95, in render
    return self.template.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 206, in render
    return self._render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 197, in _render
    return self.nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/loader_tags.py", line 173, in render
    return compiled_parent._render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 197, in _render
    return self.nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/loader_tags.py", line 173, in render
    return compiled_parent._render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 197, in _render
    return self.nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/loader_tags.py", line 173, in render
    return compiled_parent._render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 197, in _render
    return self.nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/loader_tags.py", line 69, in render
    result = block.nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/loader_tags.py", line 69, in render
    result = block.nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/loader_tags.py", line 209, in render
    return template.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 208, in render
    return self._render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 197, in _render
    return self.nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/loader_tags.py", line 173, in render
    return compiled_parent._render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 197, in _render
    return self.nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/defaulttags.py", line 326, in render
    return nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/loader_tags.py", line 69, in render
    result = block.nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/loader_tags.py", line 209, in render
    return template.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 208, in render
    return self._render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 197, in _render
    return self.nodelist.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 992, in render
    bit = node.render_annotated(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/base.py", line 959, in render_annotated
    return self.render(context)
  File "/Users/ositarska/playground/mysite/sitepackages/django/template/library.py", line 223, in render
    _dict = self.func(*resolved_args, **resolved_kwargs)
  File "/Users/ositarska/playground/mysite/sitepackages/wagtail/wagtailadmin/templatetags/wagtailadmin_tags.py", line 301, in page_listing_buttons
    for hook in button_hooks))
  File "/Users/ositarska/playground/mysite/sitepackages/wagtail/wagtailadmin/wagtail_hooks.py", line 77, in page_listing_buttons
    if page_perms.can_add_subpage():
  File "/Users/ositarska/playground/mysite/sitepackages/wagtail/wagtailcore/models.py", line 1638, in can_add_subpage
    if not self.page.specific_class.creatable_subpage_models():
  File "/Users/ositarska/playground/mysite/sitepackages/django/utils/functional.py", line 33, in __get__
    res = instance.__dict__[self.name] = self.func(instance)
  File "/Users/ositarska/playground/mysite/sitepackages/wagtail/wagtailcore/models.py", line 626, in specific_class
    content_type = ContentType.objects.get_for_id(self.content_type_id)
  File "/Users/ositarska/playground/mysite/sitepackages/djangae/contrib/contenttypes/models.py", line 102, in get_for_id
    return self.get(pk=id)
  File "/Users/ositarska/playground/mysite/sitepackages/djangae/contrib/contenttypes/models.py", line 124, in get
    dic = self._get_from_store(int(kwargs["id"]))
  File "/Users/ositarska/playground/mysite/sitepackages/djangae/contrib/contenttypes/models.py", line 113, in _get_from_store
    raise ContentType.DoesNotExist()
DoesNotExist
```

**Comment by olasitarska** [comment]
This happens here: 

```
@cached_property
    def specific_class(self):
        """
        Return the class that this page would be if instantiated in its
        most specific form
        """
        content_type = ContentType.objects.get_for_id(self.content_type_id)
        return content_type.model_class()
```

on Page class.

**Comment by erdem** [comment]
I got same error, wagtail "specific_class" page model method passing content_type object id to djangae content_type backend but djangae content_type backend expecting  "CRC64" hash key. Here is djangae stored content type data:

`{8508280596853476566L: {'app_label': 'core', 'id': 8508280596853476566L, 'model': 'megamenu'}, -3961434392430070340L: {'app_label': 'admin', 'id': -3961434392430070340L, 'model': 'logentry'}, 2298561576803913176L: {'app_label': 'core', 'id': 2298561576803913176L, 'model': 'primarynavigationitem'}, -6319035466183002790L: {'app_label': 'blog', 'id': -6319035466183002790L, 'model': 'blogarticle'}, -7768159380520649124L: {'app_label': 'core', 'id': -7768159380520649124L, 'model': 'sociallinkitem'}, -1205591047512251681L: {'app_label': 'research', 'id': -1205591047512251681L, 'model': 'promotedresearch'}, 2794123552211515624L: {'app_label': u'wagtailcore', 'id': 2794123552211515624L, 'model': 'page'}, -2339722538286455063L: {'app_label': 'taggit', 'id': -2339722538286455063L, 'model': 'taggeditem'}, -4183194102413106838L: {'app_label': u'wagtailcore', 'id': -4183194102413106838L, 'model': 'groupcollectionpermission'}, 5527819749864728082L: {'app_label': 'pages', 'id': 5527819749864728082L, 'model': 'sectorlandingtag'}}`


https://github.com/potatolondon/djangae/blob/c25c16cd1a1c738c50c75cbeb3d7b9f589463099/djangae/contrib/contenttypes/models.py#L109

In this method getting integer content type object ID.

I removed `djangae.contrib.contenttypes` from `INSTALLED_APPS` after that it's work. Is there solution for this?
